### PR TITLE
Prepare geekodoc2 release

### PR DESCRIPTION
**Pull request by tomschr**
This is a first attempt for #76. It's still an early draft and contains probably lots of inconsistency and errors.

However, I've tried to document the idea in the Makefile. Currently, it does not completely run through as it creates a circular reference errors:

```
make: Circular geekodoc/rng/5.1_1/geekodoc-v1-flat.rng <- geekodoc/rng/5.1_1/geekodoc-v1-flat.rnc dependency dropped.
make: Circular geekodoc/rng/5.1_2/geekodoc-v2-flat.rng <- geekodoc/rng/5.1_2/geekodoc-v2-flat.rnc dependency dropped.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+
+$(TARGETS): % : $(ALL_GEEKODOC)
+	@echo "Building targets: $< => $@"
+
+
+# ----------------------
+
+%.rng: %.rnc  $(ALL_ITS) $(ALL_TRANS)
+	@echo "* Converting RNG $< -> RNC $@"
```
Comment (sknorr) [inline Makefile]: Fwiw, this message exactly contradicts what this target is doing. Not sure if the message is wrong or the target definition.

**Comment by tomschr** [comment]
Thanks Stefan for your comments.

> As a general note, I think it would be much much better if you had a "build" directory instead of re/abusing the source directories for that purpose.

I'm not against a build directory (I thought of that too), but I had a reason for not to implement it. At least for the first draft.

Currently, the `develop` branch contains the flat file and the original `geekodoc5.rnc` file. In other words, *both* are versioned by git. The reason for that is, it makes it much easier for our writers: they clone this repo and can use it right away. They don't have to go through the "build dance", the requirements etc. to create the flat schema.

But maybe we can correct the Makefile so both would be possible? Something like having a `build` directory and a specific rule which copies the created flat schemas to their end location (whatever it will be)?

I don't mind where exactly the flat schema is stored in the end. But I would suggest to have the flat schema in git somewhere.

> Copying/converting all the upstream files twice for the two different Geekodoc versions seems like an unnecessary complication.

I'm pretty sure there are better ways how to do it than my current implementation. My make skills are a bit rusty. Probably @fsundermeyer has better ideas how to do that. :wink: 


> Other than that, it seems I am too dense to understand how this makefile is supposed to work because there are so many indirections here.

It's a bit complicated I know. That why I tried to document it in the header. On the other side, it's not soo different from the unmodified Makefile; the basic rules are almost the same (that's the commented part at the bottom of the file---it needs to be removed once the Makefile is ready).

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+files="$BUILD_DIR/$GEEKODOC1_PATH/$GEEKODOC1_NAME \
+       $BUILD_DIR/$GEEKODOC2_PATH/$GEEKODOC2_NAME"
+requires $files
+create_build_env $files
+rnc_to_rng $files
+make_flat $files
+cleanup_xml $files
+rngflat_to_rnc $files
+copy_flat_rnc $files
+
```
Comment (sknorr) [inline build.sh]: * A clean-up of the build dir might be positive, you only need geekodoc-flat.rn{c,g} but you get absolutely everything.

* sch-fix.xsl is not applied anywhere. Is it not necessary anymore?

* At least for Geekodoc v2, could you add a `geekodoc-its-transclusion-flat.rnc` (with ITS and Transclusion support) perhaps? (For now, maybe in addition to the regular one. But it might help to be able to dabble with those features.)

File: .gitignore [hunk 3]
```diff
@@ -15,6 +15,7 @@ tests/last-test-run-errors
 geekodoc/rng/5.1_?/*.rni
 
 build/
+.build/
```
[the remark this answers is not in the document]
Comment (tomschr) [reply .gitignore]: Yes, I've changed that now.

File: build.sh [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+BUILD_DIR=".build"
+
+# === Naming
+# The naming was quite bad, so geekodoc5 version 1 refers to DocBook5,
+# not Geekodoc v5;
+# we'll provide a compatibility link from geekodoc-v1 -> geekodoc5
+GEEKODOC1_NAME="geekodoc-v1"
+GEEKODOC2_NAME="geekodoc-v2"
+# Matches both Geekodoc names:
+GEEKODOC_REGEX_NAME="geekodoc-v?"
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: It isn't used, so I better removed it. :wink:

File: build.sh [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+GREEN="\e[35m"
+BLUE="\e[94m"
+YELLOW="\e[93m"
+CYAN="\e[36m"
+GRAY="\e[37m"
+
+BOLD="\e[1m"
+NORMAL=""
+REVERSE="\e[8m"
+RESET="\e[0m"
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: Removed them.

File: build.sh [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+function requires {
+    loginfo "Check requirements..."
+# * trang
+# * docbook_5 (from obs://Publishing)
+# * python3-rnginline (from obs://devel:languages:python/python3-rnginline)
+    local PACKAGES="trang docbook_5 python3-rnginline"
+
+    for pkg in $PACKAGES; do
+        res=$(rpm -q $pkg)
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: Yes, good point. I can check for the executables `trang` and `rnginline` instead, for example, with `type` or `which`. However, that is not possible with DocBook5.

File: build.sh [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+            exit_on_error "Package '$pkg' is missing."
+        fi
+    done
+    loginfo "All requirements are ok."
+}
+
+function create_build_env {
+    logdebug "Create build environment..."
+    # Dir structure
+    [[ -d $BUILD_DIR ]] && old ./$BUILD_DIR
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: Both are true. I think, I just remove the build directory.

File: build.sh [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    [[ -d $BUILD_DIR/$GEEKODOC1_PATH ]] || mkdir -p $BUILD_DIR/$GEEKODOC1_PATH
+    [[ -d $BUILD_DIR/$GEEKODOC2_PATH ]] || mkdir -p $BUILD_DIR/$GEEKODOC2_PATH
+
+    cp $GEEKODOC1_PATH/*.rnc $BUILD_DIR/$GEEKODOC1_PATH
+    cp $GEEKODOC2_PATH/*.rnc $BUILD_DIR/$GEEKODOC2_PATH
+    # but remove any flat files:
+    rm $BUILD_DIR/$GEEKODOC1_PATH/*-flat.rnc $BUILD_DIR/$GEEKODOC2_PATH/*-flat.rnc
+    # Copy DocBook5 schema:
+    cp $DOCBOOKXI_RNC_PATH $BUILD_DIR/$GEEKODOC1_PATH
+    cp $DOCBOOKXI_RNC_PATH $BUILD_DIR/$GEEKODOC2_PATH
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: > * A loop that iterates over both versions might be more sensible here than repeating all commands.

Unfortunately, a loop wouldn't help. The target paths are *different*.

File: build.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: I've checked DocBook with `xmlcatalog` as I had to do it anyway.

File: build.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: > * You seem to be missing something like

I've used a shell glob `*.rnc` so every RNC file is copied.

File: build.sh [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+function rnc_to_rng {
+    local files="$1"
+
+    loginfo "Convert RNC -> RNG..."
+    for f in $files; do
+       rnc="$f.rnc"
+       rng="$f.rng"
+       logdebug "Converting $rnc -> $rng"
+       trang $rnc $rng
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: Ok, fixed that.

File: build.sh (hunk 2 above)
Comment (tomschr) [reply build.sh]: > * A clean-up of the build dir at the end might be positive

I'm more in favor of doing the cleanup when the script is run for the second time. The reason for this is, sometimes I need to check/debug things in the build directory. When it's gone, that would be a bit difficult. :wink: 

Maybe we could introduce a new option for that.

> `sch-fix.xsl` is not applied anywhere. Is it not necessary anymore?

This does not affect the RNC/RNG schema. The DocBook5 schema had a bug in its Schematron patterns. This stylesheet fixes this.

> At least for Geekodoc v2, could you add a geekodoc-its-transclusion-flat.rnc [...]

I would prefer to do that in a separate PR. There are so many commits now and I would like to focus only on the building process and laying the ground for the upcoming v2 release.

Of course, if such a schema is needed, we can add it later.

File: build.sh [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    # Dir structure
+    [[ -d $BUILD_DIR ]] && old ./$BUILD_DIR
+    mkdir -p $BUILD_DIR
+    [[ -d $BUILD_DIR/$GEEKODOC1_PATH ]] || mkdir -p $BUILD_DIR/$GEEKODOC1_PATH
+    [[ -d $BUILD_DIR/$GEEKODOC2_PATH ]] || mkdir -p $BUILD_DIR/$GEEKODOC2_PATH
+
+    cp $GEEKODOC1_PATH/*.rnc $BUILD_DIR/$GEEKODOC1_PATH
+    cp $GEEKODOC2_PATH/*.rnc $BUILD_DIR/$GEEKODOC2_PATH
+    # but remove any flat files:
+    rm $BUILD_DIR/$GEEKODOC1_PATH/*-flat.rnc $BUILD_DIR/$GEEKODOC2_PATH/*-flat.rnc
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: Right. I removed the line, it doesn't make any sense.

File: build.sh [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+GEEKODOC_REGEX_NAME="geekodoc-v?"
+
+# -- Functions
+function logger() {
+    local log_priority="$1"
+    local msg="$2"
+    local color
+
+    #check if level exists
+    [[ ${LOGLEVELS[$log_priority]} ]] || return 1
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: > "return" returns from this function, meaning you never actually get to the echo at the end of it -- why not just set a default? (Same on l. 83.)

Yes, that's the point. :wink:  You only get an output if you have a certain "log level" (basically, how many `-v`s you've passed on the command line).

> In general, I think the logger functions take up a bit much space

Really? It's 15 lines (including empty lines and comments). The reals LOC is 7 lines. IMHO that's not long or "bit much space". :wink: 

> Maybe just kill the whole color support fanciness in favor of making it work..?

It was an example to have something which can be used. It works now.

**Comment by tomschr** [comment]
Thanks @sknorr, I hope I addressed all the issues you've found. Furthermore, I've reduced the commits to make it well-arranged. I used also `shellcheck` to fix the code. Very useful! :+1: :wink: 

Travis seems to be happy now.

I'm not sure what I should to with the Makefile... delete the failed attempt? Move it into a different branch? :thinking: 

Things to do:

* [ ] correct the test script
* [ ] distinguish between tests for v1 and v2 of GeekoDoc
* [ ] (re)add the test script to Travis (temporarily disabled)

**Comment by tomschr** [comment]
After looking at this PR with a little distant, I'd like to discuss the following questions:

* Are the URIs for Geekodoc useful? Currently we have:

       urn:x-suse:rng:geekodoc5.rnc

  However, this URI should be considered as deprecated and replaced with:

      urn:x-suse:rnc:v1:geekodoc

  Likewise for GeekoDoc v2. The above doesn't take into account the DocBook version.
  Not sure if we really need that? Maybe something with:

      urn:x-suse:rnc:db51:v1:geekodoc

* Do you think the directories `5.1_1` and `5.1_2` are a good idea? They indicate the
  DocBook version. Or should this be part of the filename?

File: tests/v2/bad/wrong-xmlid-value.xml [hunk 12]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?xml-model href="file:/home/toms/repos/GH/opensuse/geekodoc/geekodoc/rng/5.1_1/xmlid-test.rnc" type="application/relax-ng-compact-syntax"?>
```
Comment (sknorr) [inline tests/v2/bad/wrong-xmlid-value.xml:2]: Not sure if you really want the name of your home dir in there.

**Review by sknorr — CHANGES_REQUESTED** [review]
So, the build script overall appears to work, even though it is a bit confusing. I have not checked any content against the new GeekoDoc version at all (yet).

The repo content is now extremely confusing however, everything appears to be copied/symlinked everywhere and there is no clear sense which directories are input directories and which are output directories. There are two "tests" dirs (tests/ and geekodoc/tests/). The generated flat schemas are still part of the source dirs instead of being confined to the build dir. It would also be good to have a dir called e.g. "upstream" to collect DocBook, ITS, Transclusion, RelaxNG schemas (ideally with clearly versioned file names), rather than having them as part of the "geekodoc" dir. I am also not sure why you're using the system DocBook schema when the repo itself is already shipping a fixed version the same schema (dbits.rng on the other hand is not part of this repo). You also continue to include schema-fix.xsl in the repo, even though it is no longer needed apparently.

**Comment by tomschr** [comment]
Thanks Stefan for all your efforts! Much appreciated. :+1: 
Hope I can answer some of your concerns. :slightly_smiling_face: 

> The repo content is now extremely confusing however, everything appears to be copied/symlinked everywhere and there is no clear sense which directories are input directories and which are output directories. 

I'm sorry, but I don't know how this could be solved. It's hard to _not_ symlink if you need to share files in both versions, be it inside the schema or in the `tests` dir. I don't want to repeat myself.
Any ideas?

* Input directory: `geekodoc`
* Output directory: `build/`

> There are two "tests" dirs (tests/ and geekodoc/tests/).

That is solved now. The test in `geekodoc/tests` was a leftover, I removed it.


> The generated flat schemas are still part of the source dirs instead of being confined to the build dir.

The schema is created in the `build` directory.

But yes, it's also copied into the source dirs on purpose. The reason for this is, that everybody can clone or update the repo and have the file "at hand" (without going through the build process). I thought it is as a "good thing" _in this case_. (I'm aware that the built results shouldn't be committed into a VCS.)

Not sure if we could change that. That would affect our users as they have to build the schema manually. 
Maybe, as a compromise, we could use a different directory apart from the source dir where we store the outcome of the build. Would that work?

>  It would also be good to have a dir called e.g. "upstream" to collect DocBook, ITS, Transclusion, RelaxNG schemas [...]

Although I like the idea, that would make editing with oXygen more difficult, unfortunately.

> I am also not sure why you're using the system DocBook schema when the repo itself is already shipping a fixed version the same schema (dbits.rng on the other hand is not part of this repo).

I used it to make editing with oXygen possible. Although I have to admit, maybe that can be changed. (I don't know how ATM).

> You also continue to include schema-fix.xsl in the repo, even though it is no longer needed apparently.

Right, removed it.

----

Will cover the individual changes later.

**Comment by sknorr** [comment]
> I'm sorry, but I don't know how this could be solved. It's hard to _not_ symlink if you need to share files in both versions, be it inside the schema or in the `tests` dir. I don't want to repeat myself.
> [snipped]
>     * Input directory: `geekodoc`
>     * Output directory: `build/`

Yes, that's what I'd have hoped, except that's not really the case currently, as `geekodoc` is used for both input and output. 
The schema is also linked into the I don't really understand why the schema needs to be linked into the `tests` dir for example. Why can't you just test the schema that is generated into the `build` dir?

If I can make a suggestion, then I'd say:

* `geekodoc/` - GeekoDoc schema _input_ only
* `external/` - Everything that comes from upstream (ITS, DocBook, RelaxNG, ...)
* `tests/` - Test cases as input
* `build/intermediate` - Intermediate files that won't be of further use
* `build/geekodoc` - Output tree that can be packaged as-is
* `build/tests/` in case there are any files that need to be written by a test case (optional)

You could put generated schemas on the `gh-pages` branch (or have them auto-generated there via GitHub Actions), that way they'd not be part of the normal repo and they'd have an official spot. But ultimately, I don't think having them as part of the repo is not strictly necessary. The main way we ship the schema is the RPM, everyone else can just install the necessary dependencies to build the schema.
(Which should also answer some of your questions further down in the comment.)


> That is solved now. The test in `geekodoc/tests` was a leftover, I removed it.

Good :)

> > It would also be good to have a dir called e.g. "upstream" to collect DocBook, ITS, Transclusion, RelaxNG schemas [...]
> Although I like the idea, that would make editing with oXygen more difficult, unfortunately.

What is the issue with Oxygen? Is it somehow allergic to situations where schemas lie in different directories?
I think I actually don't understand your schema editing workflow at all. I figure you're not doing it the way I do it in a bare text editor then?

> > I am also not sure why you're using the system DocBook schema when the repo itself is already shipping a fixed version the same schema (dbits.rng on the other hand is not part of this repo).
> 
> I used it to make editing with oXygen possible. Although I have to admit, maybe that can be changed. (I don't know how ATM).

To be honest, I think there's a certain charm to including all the necessary schema dependencies in the repo. And clearly, you're doing it for other upstream schemas already, so why not also for all the necessary parts of DocBook? That would also avoid having to check for the system path of the DocBook schema in the build script.

> > You also continue to include schema-fix.xsl in the repo, even though it is no longer needed apparently.
> Right, removed it.

Thanks!

**Comment by tomschr** [comment]
Thanks Stefan for your thoughts. :+1: 

> [snipped]
> If I can make a suggestion, then I'd say:
>
>   `geekodoc/` - GeekoDoc schema; input files only
>   `external/` - Everything that comes from upstream (ITS, DocBook, RelaxNG, ...)
>   `tests/` - Test cases; input files only
>    `build/intermediate` - Intermediate files that won't be of further use
>    `build/geekodoc` - Output tree that can be packaged as-is (i.e. no extraneous files that won't be needed anymore after that)
>    `build/tests/` in case there are any files that need to be written by a test case (optional)

The `build/geekodoc` might be useful, not sure about `build/intermediate`. These files are not many. But I need to code it to see if this distinction would be helpful.

> You could put generated schemas on the gh-pages branch (or have them auto-generated there via GitHub Actions)

I like the GitHub Action idea! Either we could store the schema on gh-pages or on  releases. I will look into that.

> But ultimately, I don't think having them as part of the repo is not strictly necessary. The main way we ship the schema is the RPM, everyone else can just install the necessary dependencies to build the schema.

It was just a lazy and convenient way to tell people "if you need the latest Geekodoc schema, clone the repo and your basically done."
If we don't want to store the schema files in the repo, we need to look for a replacement. You suggested RPM or with a GH Action idea. I think, both have their merits.

To be useful, we need to use a service file to automatically build the RPM. My idea would be to store the `geekodoc` package in `Documentation:Tools:Auto`. That would trigger a rebuild on every commit. So this package is considered as "use at your own risk".

After development is finished, I could submit this package to D:T to mark it as stable. That part needs to be manual. Does that make sense?


> What is the issue with Oxygen? Is it somehow allergic to situations where schemas lie in different directories?

Well, to be precise, _any_ tool would be allergic to different directories. OXygen was just a specific example to make my statement a bit simpler. :wink: The problem here is not oXygen per se, it's the `include "docbook.rnc"` statement in the RNC file.

If I load the GeekoDoc schema in oXygen, the editor tries to load the DocBook schema from the same place. If I move it to another directory, oXygen will fail. That will give me errors. Maybe that explains why I used symlinks etc. so much. :smile: 

That brings me another idea. Maybe I could use a XML catalog, local to that repo... :thinking:
The only drawback that I don't like is: it adds another level of complexity. You can't just load the GeekoDoc RNC and start editing, you need to configure oXygen a bit. It's certainly not a big deal, but still... Probably that could be solved by project files.

> I think I actually don't understand your schema editing workflow at all. I figure you're not doing it the way I do it in a bare text editor then?

It's actually nothing special. Just as described in my last paragraph.
Sure, I could ignore the error, but then oXygen doesn't give me all the benefits. Would be the same like driving a car with the hand brake on. :smile: 

> To be honest, I think there's a certain charm to including all the necessary schema dependencies in the repo. And clearly, you're doing it for other upstream schemas already, so why not also for all the necessary parts of DocBook? That would also avoid having to check for the system path of the DocBook schema in the build script.

Yes, I was thinking about that. Maybe I should include the DocBook schemas here to remove a dependency.

I'm aware, that my approach was a bit "dirty" and your suggestion would make the repo cleaner.

As a lot of stuff is involved here, maybe it could make sense to discuss some parts in our Wednesday tools meeting with Frank.

----

TL;DR: So to summarize the above findings:

* Look into a cleaner repo structure to separate input and output directories.
* Make use of service file to build RPM automatically and store it in D:T:Auto
* Make use of GitHub Actions to store GeekoDoc schemas in gh-pages or under releases.
* Create an oXygen project file to load local XML catalog files pointing to the right locations
* Store DocBook schema files inside this repo.


Thanks!

**Comment by tomschr** [comment]
> Create an oXygen project file to load local XML catalog files pointing to the right locations

Did a quick check and I can report success! :tada: 

Seems it's possible. Will adding a `.xpr` file (oXygen project file) and a XML catalog file soon. I think, that's a small price to pay to make the repo cleaner.

**Comment by tomschr** [comment]
Short interim result:

> Look into a cleaner repo structure to separate input and output directories.

I think, I can report another progress. As you suggested, the DocBook schemas are now stored under `external/docbook5` separated from the GeekoDoc sources.

Once the `build.sh` script is finished, the result is stored under a `dist/` directory (for distribution). I copied that from the Python way. :wink: I think that should do it as long it is not stored inside the source dirs.

Continue tomorrow...

File: README.md [hunk 13]
```diff
@@ -8,16 +8,59 @@
...
+
+Currently, to support legacy documentation, we split the GeekoDoc schema into
+two versions:
+
+* **Version 1** this is the legacy version.
+* **Version 2** this is where all the new features are added.
+
+The two versions can be accessed by its URI.
+
+* Version 1 URIs
```
[the remark this answers is not in the document]
Comment (tomschr) [reply README.md:30]: > This is _really_ confusing because there are now v1 and v2 URIs and v1 and v2 Geekodoc.

Well, if we have two GeekoDoc versions we need also to distinguish them, right?

> In addition, I always used this as the URI value in DAPS: DOCBOOK5_RNG_URI="https://github.com/openSUSE/geekodoc/raw/master/geekodoc/rng/geekodoc5-flat.rnc". This style is not documented at all here ...

Yeah, added a separate section about _Using GeekoDoc with DAPS_.

You can add any URI into the env variable as long as it can be resolved though our (GeekoDoc) XML catalog.

File: build.sh [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+SYNOPSIS
+  $ME -h|--help
+  $ME [OPTIONS]
+
+OPTIONS
+  -v, -vv, -vvv   Log level, the more you add, the more messages
+                  you get. Restricted to -vvv (=DEBUG)
+  -b DIR, --builddir=DIR
+                  Set the directory where to build (default: ".build")
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: That was fixed in one of the latest commits. Actually I used the variable `BUILD_DIR` for that.

File: build.sh [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+    loginfo "Check requirements..."
+# * trang
+# * python3-rnginline (from obs://devel:languages:python/python3-rnginline)
+# * docbook_5 (from obs://Publishing)
+    local SCRIPTS="trang rnginline"
+    local res
+
+    for script in $SCRIPTS; do
+        res=$(command -v "$script" )
+        if [ 1 -eq "$?" ]; then
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh:148]: Hmn, I did that for a purpose but I cannot find that explanation anymore. Anyway, I use double brackets here which should solve the issue.

File: build.sh [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+function get_docbook5_path {
+    local path
+    local res
+    loginfo "Retrieve XML catalog entry..."
+
+    path=$(xmlcatalog /etc/xml/catalog "$DOCBOOKXI_URI")
+    res=$?
+    logdebug "Detected OS '$ID'..."
+
+    if [ 0 -gt "$res" ]; then
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: That part isn't included anymore and was removed in commit 197ce8f6c.

File: build.sh [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+        # success, use the result, but cut off the "file://" schema
+        path="${path#file://*}"
+    fi
+    logdebug "Use '$path' path for DocBook5"
+}
+
+function create_build_env {
+    loginfo "Create build environment..."
+    # Dir structure
+    if [[ -n $BUILD_DIR ]] && [[ -d $BUILD_DIR ]]; then
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh:160]: > Why are you checking whether the BUILD_DIR variable exists here? 

Actually I'm checking for a non-empty string. But I moved that check now when checking the `-b`/`--builddir` option.

File: build.sh [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+VERBOSITY=0
+LOGGING_LEVEL="DEBUG"
+declare -A LOGLEVELS=([DEBUG]=0 [INFO]=1 [WARN]=2 [ERROR]=3)
+declare -A LEVEL2LOG=([0]="ERROR" [1]="WARN" [2]="INFO" [3]="DEBUG")
+declare -A LOGCOLORS=([DEBUG]=$CYAN [INFO]=$BOLD [WARN]=$YELLOW [ERROR]=$RED)
+
+# -- Paths
+DOCBOOKXI_URI="http://www.oasis-open.org/docbook/xml/5.1/rng/docbookxi.rnc"
+DOCBOOKXI_RNC_PATH="/usr/share/xml/docbook/schema/rng/5.1/docbookxi.rnc"
+DBITS_RNC_PATH="/usr/share/xml/docbook/schema/rng/5.1/dbits.rnc"
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: I removed that part in commit 197ce8f as we have the `external` directory.

File: build.sh [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+    cp $GEEKODOC1_PATH/*.rnc $BUILD_DIR/$GEEKODOC1_PATH
+    cp $GEEKODOC2_PATH/*.rnc $BUILD_DIR/$GEEKODOC2_PATH
+    # Copy DocBook5 schema for GeekoDoc v1
+    cp $DOCBOOKXI_RNC_PATH $BUILD_DIR/$GEEKODOC1_PATH
+
+    # Copy DocBook5 schema and ITS for GeekoDoc v2
+    cp $DOCBOOKXI_RNC_PATH $BUILD_DIR/$GEEKODOC2_PATH
+    pushd $BUILD_DIR/$GEEKODOC2_PATH
+    ln -s $DBITS_RNC_PATH
+    # cp $DBITS_RNC_PATH $BUILD_DIR/$GEEKODOC2_PATH
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: It's removed in 197ce8f.

File: build.sh [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+
+function create_build_env {
+    loginfo "Create build environment..."
+    # Dir structure
+    if [[ -n $BUILD_DIR ]] && [[ -d $BUILD_DIR ]]; then
+       rm -rf $BUILD_DIR 2>/dev/null
+    fi
+    mkdir -p $BUILD_DIR
+    [[ -d $BUILD_DIR/$GEEKODOC1_PATH ]] || mkdir -p $BUILD_DIR/$GEEKODOC1_PATH
+    [[ -d $BUILD_DIR/$GEEKODOC2_PATH ]] || mkdir -p $BUILD_DIR/$GEEKODOC2_PATH
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh:166]: Hmn, seems to be a leftover. Removed.

File: build.sh [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+    if [ 0 -gt "$res" ]; then
+        # error, could not be found, so fall back
+        case $ID in
+          opensuse*|sle*)
+            path="/usr/share/xml/docbook/schema/rng/5.1/docbookxi.rnc"
+            ;;
+          debian*|ubuntu*)
+            path="/usr/share/xml/docbook/schema/rng/5.1/docbookxi.rnc"
+            ;;
+          fedora|redhat*)
```
[the remark this answers is not in the document]
Comment (tomschr) [reply build.sh]: As we have an external directory which holds all the schemas, this is not needed anymore. Problem solved. :wink:

File: tests/v2/bad/wrong-xmlid-value.xml (hunk 12 above)
Comment (tomschr) [reply tests/v2/bad/wrong-xmlid-value.xml:2]: Yes, was a accident and shouldn't be committed. Removed in 197ce8f.

File: tests/run-tests.sh [hunk 22]
```diff
@@ -8,23 +8,32 @@
...
+
+# SCHEMA=${PROGDIR}/../rng/geekodoc5-flat.rng
+# SCHEMA=$(readlink -f ${SCHEMA})
 ERRORS=0
-ERRORFILE=${PROGDIR}/last-test-run-errors
+ERRORFILE="${PROGDIR}/last-test-run-errors"
+
+# Which versions of Geekodoc should be tested?
+# valid: 1, 2, both
+GEEKODOC="both"
```
[the remark this answers is not in the document]
Comment (tomschr) [reply tests/run-tests.sh]: Good idea! Integrated in commit 5a77933.
